File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/util/ClassUtils.java
package org.javamaster.invocationlab.admin.util;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * @author yudong
 * @date 2022/11/12
 */
public class ClassUtils {

    public static Type getParamGenericType(Type type) {
        if (type instanceof ParameterizedType) {
            ParameterizedType parameterizedType = (ParameterizedType) type;
            Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
            return actualTypeArguments[0];
        }
        return null;
    }

    public static boolean shouldNew(Type type) {
        if (type == null) {
            return false;
        }
        String typeName = type.getTypeName();
        return shouldNew(typeName);
    }

    public static boolean shouldNew(String typeName) {
        return !isPrimitive(typeName) && !typeName.startsWith("java");
    }

    public static boolean isPrimitive(String typeName) {
        switch (typeName) {
            case "int":
            case "char":
            case "long":
            case "boolean":
            case "float":
            case "double":
            case "[B":
            case "byte[]":
            case "java.lang.String":
            case "java.lang.Integer":
            case "java.lang.Long":
            case "java.lang.Double":
            case "java.lang.Float":
            case "java.math.BigInteger":
            case "java.math.BigDecimal":
                return true;
            default:
                return false;
        }
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/util/ClassUtils.kt
package org.javamaster.invocationlab.admin.util

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * @author yudong
 * @date 2022/11/12
 */
object ClassUtils {
    @JvmStatic
    fun getParamGenericType(type: Type?): Type? {
        if (type is ParameterizedType) {
            val actualTypeArguments = type.actualTypeArguments
            return actualTypeArguments[0]
        }
        return null
    }

    @JvmStatic
    fun shouldNew(type: Type?): Boolean {
        if (type == null) {
            return false
        }
        val typeName = type.typeName
        return shouldNew(typeName)
    }

    private fun shouldNew(typeName: String): Boolean {
        return !isPrimitive(typeName) && !typeName.startsWith("java")
    }

    @JvmStatic
    fun isPrimitive(typeName: String?): Boolean {
        return when (typeName) {
            "int", "char", "long", "boolean", "float", "double", "[B", "byte[]", "java.lang.String", "java.lang.Integer", "java.lang.Long", "java.lang.Double", "java.lang.Float", "java.math.BigInteger", "java.math.BigDecimal" -> true
            else -> false
        }
    }
}
